### PR TITLE
jobsdb tests

### PR DESCRIPTION
**Pull request by Sidddddarth**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by chandumlg — APPROVED** [review]

File: jobsdb/jobsdb_test.go [hunk 1]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+	EventPayload: []byte(`{"receivedAt":"2021-06-06T20:26:39.598+05:30","writeKey":"1rNMpxFxVdoaAdItcXTbVVWdonD","requestIP":"[::1]",  "batch": [{"anonymousId":"anon_id","channel":"android-sdk","context":{"app":{"build":"1","name":"RudderAndroidClient","namespace":"com.rudderlabs.android.sdk","version":"1.0"},"device":{"id":"49e4bdd1c280bc00","manufacturer":"Google","model":"Android SDK built for x86","name":"generic_x86"},"library":{"name":"com.rudderstack.android.sdk.core"},"locale":"en-US","network":{"carrier":"Android"},"screen":{"density":420,"height":1794,"width":1080},"traits":{"anonymousId":"49e4bdd1c280bc00"},"user_agent":"Dalvik/2.1.0 (Linux; U; Android 9; Android SDK built for x86 Build/PSR1.180720.075)"},"event":"Demo Track","integrations":{"All":true},"messageId":"b96f3d8a-7c26-4329-9671-4e3202f42f15","originalTimestamp":"2019-08-12T05:08:30.909Z","properties":{"category":"Demo Category","floatVal":4.501,"label":"Demo Label","testArray":[{"id":"elem1","value":"e1"},{"id":"elem2","value":"e2"}],"testMap":{"t1":"a","t2":4},"value":5},"rudderId":"90ca6da0-292e-4e79-9880-f8009e0ae4a3","sentAt":"2019-08-12T05:08:30.909Z","type":"track"}]}`),
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:186]: add a fake write key here?

File: jobsdb/jobsdb_test.go [hunk 2]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+
+			ds = jd.datasetList[len(jd.datasetList)-1]
+		})
+
+		It("should store jobs to db through workers", func() {
+			jd.enableWriterQueue = true
+
+			c.mock.ExpectBegin()
+			stmt := c.mock.ExpectPrepare(fmt.Sprintf(`COPY "%s" ("uuid", "user_id", "custom_val", "parameters", "event_payload") FROM STDIN`, ds.JobTable))
+			for _, job := range properStoreJobs {
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:282]: properStoreJobs to mockedStoreJobs?

File: jobsdb/jobsdb_test.go [hunk 3]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+		stats.Setup()
+
+		globalDBHandle = c.globalDB
+	})
+
+	AfterEach(func() {
+		c.Finish()
+	})
+
+	Context("getDSList unit test", func() {
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:226]: rename it to just the function name without 'unit test'

File: jobsdb/jobsdb_test.go [hunk 4]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+	Context("UpdateJobStatus", func() {
+		var jd *HandleT
+
+		BeforeEach(func() {
+			jd = &HandleT{}
+			jd.dbHandle = c.db
+			jd.datasetList = dsListInMemory
+			jd.datasetRangeList = dsRangeList
+			jd.enableWriterQueue = true
+			jd.workersAndAuxSetup(ReadWrite, "tt", 0*time.Hour, "", false, QueryFiltersT{})
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:406]: rename tt to some-prefix

File: jobsdb/jobsdb_test.go [hunk 5]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+	Context("GetProcesed", func() {
+		var jd *HandleT
+
+		BeforeEach(func() {
+			jd = &HandleT{}
+			jd.dbHandle = c.db
+			jd.datasetList = dsListInMemory
+			jd.datasetRangeList = dsRangeList
+			jd.enableWriterQueue = true
+			jd.workersAndAuxSetup(ReadWrite, "tt", 0*time.Hour, "", false, QueryFiltersT{})
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:479]: rename tt to some-prefix

File: jobsdb/jobsdb_test.go [hunk 6]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+		}
+
+		nonTerminalStates := []string{"executing", "failed", "waiting", "throttled"}
+		for _, nonTerminalState := range nonTerminalStates {
+			assertGetProcessedJobsWithCustomVal(nonTerminalState)
+		}
+
+		assertGetProcessedJobsWithParameters := func(state string) {
+			It("should return jobs with non terminal last state when queried with parameters", func() {
+				destinationID := "did1"
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:526]: rename did1 to some-destination-d

File: jobsdb/jobsdb_test.go [hunk 7]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+			jd.datasetRangeList = dsRangeList
+			jd.enableWriterQueue = true
+			jd.workersAndAuxSetup(ReadWrite, "tt", 0*time.Hour, "", false, QueryFiltersT{})
+		})
+
+		assertGetProcessedJobsWithCustomVal := func(state string) {
+			It("should return jobs with non terminal last state when queried with customval", func() {
+				var stateQuery, customValQuery, sourceQuery, limitQuery string
+				stateQuery = " AND ((job_state='failed'))"
+				customValQuery = " AND ((tt_jobs_1.custom_val='GW'))"
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:486]: Do we need to hardcode the custom val to 'GW' or can we something like 'some-ds-name'

File: jobsdb/jobsdb_test.go [hunk 8]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+		It("should delete only one executing with simple customVal", func() {
+			timeNow := time.Now()
+			getTimeNowFunc = func() time.Time {
+				return timeNow
+			}
+
+			c.mock.ExpectBegin()
+
+			ds := dsListInMemory[0]
+			customValQuery := "tt_jobs_1.custom_val='WEBHOOK'"
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:694]: would it work if we rename WEBHOOK to some-destination-name

File: jobsdb/jobsdb_test.go [hunk 9]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+
+var mockRows = func() *sqlmock.Rows {
+	sqlMockRows := sqlmock.NewRows([]string{"tablename"})
+	for _, row := range tablesNamesInDB {
+		sqlMockRows.AddRow(row)
+	}
+	return sqlMockRows
+}()
+
+var testJobs = []*JobT{
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:894]: rename to mockJobs and remove original SourceID's

File: jobsdb/jobsdb_test.go [hunk 10]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+			break
+		}
+		sqlMockRows.AddRow(job.JobID, job.UUID, job.UserID, job.Parameters, job.CustomVal,
+			job.EventPayload, job.CreatedAt, job.ExpireAt, job.LastJobStatus.JobState, job.LastJobStatus.AttemptNum,
+			job.LastJobStatus.ExecTime, job.LastJobStatus.RetryTime, job.LastJobStatus.ErrorCode, job.LastJobStatus.ErrorResponse)
+	}
+	return sqlMockRows
+}
+
+var properStoreJobs = []*JobT{
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:976]: Remove original source-id's and writekeys

File: jobsdb/jobsdb_test.go [hunk 11]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+		EventPayload: []byte(`{}`),
+		UserID:       "",
+		UUID:         uuid.NewV4(),
+		CustomVal:    "WEBHOOK",
+	},
+}
+
+var uuidStr = "a362c501-c38e-4aee-ae61-4a3b095ebcab"
+var s, _ = uuid.FromString(uuidStr)
+var partiallyProperStoreJobs = []*JobT{
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:1002]: rename this to mockedPartiallyStoredJobs

File: jobsdb/jobsdb_test.go [hunk 12]
```diff
@@ -169,3 +180,859 @@ var _ = Describe("Calculate newDSIdx for cluster migrations", func() {
...
+
+			job1 := partiallyProperStoreJobs[0]
+			job2 := partiallyProperStoreJobs[1]
+			stmt := c.mock.ExpectPrepare(fmt.Sprintf(`INSERT INTO %s (uuid, user_id, custom_val, parameters, event_payload)
+			VALUES ($1, $2, $3, $4, (regexp_replace($5::text, '\\u0000', '', 'g'))::json) RETURNING job_id`, ds.JobTable))
+			stmt.ExpectExec().WithArgs(job1.UUID, job1.UserID, job1.CustomVal, string(job1.Parameters), string(job1.EventPayload)).WillReturnResult(sqlmock.NewResult(0, 1))
+			stmt = c.mock.ExpectPrepare(fmt.Sprintf(`INSERT INTO %s (uuid, user_id, custom_val, parameters, event_payload)
+			VALUES ($1, $2, $3, $4, (regexp_replace($5::text, '\\u0000', '', 'g'))::json) RETURNING job_id`, ds.JobTable))
+			err := &pq.Error{}
+			err.Code = "22P02"
```
Comment (cisse21) [inline jobsdb/jobsdb_test.go:384]: Comment indicating what the error code is

**Review by cisse21 — CHANGES_REQUESTED** [review]

**Review by cisse21 — APPROVED** [review]

**Pull request merged**